components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -760,43 +773,52 @@ internal sealed class SubscriptionWrapper : ISubscription {
-                                    if (heartbeatValue != null) {
-                                        heartbeatValue.SequenceNumber = sequenceNumber;
-                                        heartbeatValue.IsHeartbeat = true;
-                                        heartbeatValue.PublishTime = publishTime;
-                                        if (message.Notifications == null) {
-                                            message.Notifications =
-                                                new List<MonitoredItemNotificationModel>();
-                                        }
-                                        message.Notifications.Add(heartbeatValue);
-                                    }
-                                    continue;
+                                Exists(m => m.ClientHandle == item.Item.ClientHandle)
+                                && item.ValidateHeartbeat(publishTime)) {
+
+                                MonitoredItemNotificationModel GetDefaultNotification()
+                                {
+                                    return new MonitoredItemNotificationModel {
+                                        Id = item?.Template?.Id,
+                                        DisplayName = item.Item.DisplayName,
+                                        NodeId = item?.Template?.StartNodeId,
+                                        AttributeId = item.Item.AttributeId,
+                                        ClientHandle = item.Item.ClientHandle,
+                                        Value = new DataValue(Variant.Null,
+                                            item.Item?.Status?.Error?.StatusCode ??
+                                            StatusCodes.BadMonitoredItemIdInvalid),
+                                        Overflow = false,
+                                        NotificationData = null,
+                                        StringTable = null,
+                                        DiagnosticInfo = null,
+                                    };
                                 }
+
+                                var heartbeatValue = item.Item?.LastValue.
+                                    ToMonitoredItemNotification(item.Item, GetDefaultNotification);
+                                if (heartbeatValue != null) {
+                                    heartbeatValue.SequenceNumber = sequenceNumber;
+                                    heartbeatValue.IsHeartbeat = true;
+                                    heartbeatValue.PublishTime = publishTime;
+                                    message.Notifications?.Add(heartbeatValue);
+                                }
+                                continue;
                             }
                             item.ValidateHeartbeat(publishTime);
                         }
                     }
 
+                    var erroneousNotifications = message.Notifications?
```

Comment:
Since this is costly, maybe hide it behind a check for Debug Level being enabled on the logger?

Reply:
Moved calculation behind a check.